src/plugins/intellisense/index.tsx

```diff
@@ -0,0 +1,512 @@
+  return Object.keys(Calc.controller.listModel.selectedItemMap)[0];
+}
+
+export default class Intellisense extends PluginController {
+  static id = "intellisense" as const;
+  static enabledByDefault = true;
+
+  view: MountedComponent | undefined;
+
+  x: number = 0;
+  y: number = 0;
+
+  intellisenseOpts: BoundIdentifier[] = [];
+  intellisenseIndex: number = -1;
+
+  latestIdent: TryFindMQIdentResult | undefined;
+  latestMQ: MathQuillField | undefined;
+
+  intellisenseReturnMQ: MathQuillField | undefined;
+  prevCursorPos: { x: number; y: number } | undefined;
+  goRightBeforeReturningToMQ: boolean = false;
+
+  idcounter = 0;
+
+  partialFunctionCall: PartialFunctionCall | undefined;
+  partialFunctionCallIdent: BoundIdentifier | undefined;
+  partialFunctionCallDoc: string | undefined;
+
+  intellisenseState = new IntellisenseState(getMetadata());
+
+  canHaveIntellisense = false;
+
+  // recalculate the intellisense
+  updateIntellisense() {
+    const focusedMQ = MathQuillView.getFocusedMathquill();
+    this.saveCursorState();
+    this.intellisenseOpts = [];
+
+    // is there actually a focused mathquill window?
+    if (focusedMQ) {
+      // find the identifier the cursor is at
+      this.latestIdent = getMathquillIdentifierAtCursorPosition(focusedMQ);
+      if (this.latestIdent)
+        this.latestIdent.ident = this.latestIdent.ident.replace(/ /g, "");
+
+      this.latestMQ = focusedMQ;
+
+      // determine if the user is in a partial function call
+      this.partialFunctionCall = getPartialFunctionCall(focusedMQ);
+      this.partialFunctionCallIdent = Array.from(
+        this.intellisenseState.boundIdentifiers()
+      ).find(
+        (i) =>
+          addBracketsToIdent(i.variableName) ===
+            this.partialFunctionCall?.ident && i.type === "function"
+      );
+
+      // if the user is in a partial function call,
+      // find its documentation if it exists
+      const models = Calc.controller.getAllItemModels();
+      let found = false;
+      for (let i = 0; i < models.length; i++) {
+        const current = models[i];
+        if (
+          this.partialFunctionCallIdent &&
+          current.type === "text" &&
+          models[i + 1]?.type === "expression" &&
+          this.partialFunctionCallIdent.exprId === models[i + 1]?.id
+        ) {
+          this.partialFunctionCallDoc = current.text;
+          found = true;
+        }
+      }
+      if (!found) this.partialFunctionCallDoc = undefined;
+
+      // determine where to put intellisense window
+      const bbox = getMQCursorPosition(focusedMQ);
+      this.x = bbox?.left ?? 0;
+      this.y = bbox?.top ?? 0;
+
+      // create filtered list of valid intellisense options
+      if (this.latestIdent) {
+        this.intellisenseOpts = Array.from(
+          this.intellisenseState.boundIdentifiers()
+        ).filter((g) =>
+          g.variableName.startsWith(
+            this.latestIdent?.ident.replace(/[{} \\]/g, "") ?? ""
```

Comment:
I only see this regex `/[{} \\]/g` appear once. How does `g.variableName` get correspondingly cleaned up?

Reply:
The reason why I have to use the cleanup regex `/[{} \\]/g` is because `this.latestIdent.ident` is generated from regular LaTeX, whereas `g.variableName` is generated from an aug expression, which seem to have different formats.

Reply:
Why not parse `this.latestIdent.ident` to Aug, and compare then? (maybe performance). According to the comment below, Aug identifiers are pretty normalized. (I don't know what I'm saying about curly braces there). (looks like we're relying on Desmos to do that normalization? see `Desmos.Private.Parser.parse("\\operatorname{theta}_{sdf}")._symbol`)

Some brutal test cases: https://www.desmos.com/calculator/nbyl2hhzyh. params for h() doesn't show at all, and i() doesn't show usage.

https://github.com/DesModder/DesModder/blob/7a76fcd45d89a31ef7dced24239f7140cdd410fa/src/plugins/text-mode/aug/AugLatex.ts#L97-L108

Reply:
Good idea. I'll switch to Aug identifiers for everything.

Reply:
Fixed (non-latex identifiers are used as a default now)